ports/esp8266/modules/websocket_helper: Improve client handshake's compliance with RFC6455.

ccooper21 opened the pull request:
The existing client handshake implementation is non-compliant with RFC6455
Sections 1.3 and 4.1, and is only sufficient for use with the existing server
handshake implementation.  This change increases the client handshake's
compliance with the RFC by fully supporting the "Sec-WebSocket-Key",
"Sec-WebSocket-Protocol", "Sec-WebSocket-Version", and "Sec-WebSocket-Accept"
HTTP headers.  The value of the "Sec-WebSocket-Protocol" response header is
returned to the caller for further handling.  To minimize the code footprint,
if a server returns a "Sec-WebSocket-Accept" response header and it has the
expected value, it is assumed that the HTTP connection has been successfully
upgraded to a websocket connection.  Other aspects of the HTTP response are
not validated (i.e. the response code, the "Upgrade" header, and the
"Connection" header).

ccooper21 commented:
Here is some additional information regarding this pull request:

- I successfully tested the enhanced client handshake with the AWS IoT MQTT-based service and HiveMQ MQTT service.  I also confirmed that the ESP8266 `webrepl` functionality is unaffected by this change.

- Since the client handshake depends on random key generated using the `urandom` module, it is a best practice to first seed the random number generator.  I'm not aware of any existing documentation for the websocket related modules, so I'm not sure where to make note of this other than here.

- I eliminated the `usocket.makefile(...)` call in the client handshake function to provide compatability with the `ussl` module.  This call appears to have been a no-op on the ESP8266 platform anyway.

- While it pained me, I used extremely short variable names to minimize the code footprint (e.g. "sps" = "subprotocols", "dhs" = "desired_headers", "dvs" = "desired_values").  Some of this pain comes from seeing the existing short variable names for which I don't even have a guess as to their meaning (e.g. "cl", "clr"), and contributing to this further.  The exception to this are the names of the helper functions that I added.  If you have a preference on the names of these, let me know.

As per the contribution guidelines, I attemped to follow the conventions established in the changed file and avoided making changes unrelated to this pull request's goal.  Given this, I also recommend making the following changes.  If there is interest in me proceeding, let me know and I can make another pull request or add them to this one.

- Eliminate the `usocket.makefile(...)` call in the server handshake function for consistency with the client handshake, and as a step towards enabling a websocket server's use of SSL/TLS.

- Replace the use of block quotes with implied inline continuation.  Block quotes are presently used for the static HTTP content.  With the multiple chunks of HTTP content now in use, I find this hard to read since the block quotes necessarily don't preserve the indentation of the code.  Also, given that HTTP requires each line to end with a `\r\n`, the use of an explicit `\r` followed by an implicit `\n` may cause confusion when maintaing the code.

- Some debug messages refer to HTTP header names while others refer to variable names.  This should be reconciled.

pfalcon commented:
Can you please start with saying what the problem this patch fixes? (Extra points for starting the description with that info.)

ccooper21 commented:
For test scripts related to this pull request, please see [here](https://gist.github.com/ccooper21/1a4a65cb7d850ac6872764e838993b0e).

ccooper21 committed:
ports/esp8266/modules/websocket_helper: Improve client handshake's compliance with RFC6455.

The existing client handshake implementation is non-compliant with RFC6455
Sections 1.3 and 4.1, and is only sufficient for use with the existing server
handshake implementation.  This change increases the client handshake's
compliance with the RFC by fully supporting the "Sec-WebSocket-Key",
"Sec-WebSocket-Protocol", "Sec-WebSocket-Version", and "Sec-WebSocket-Accept"
HTTP headers.  The value of the "Sec-WebSocket-Protocol" response header is
returned to the caller for further handling.  To minimize the code footprint,
if a server returns a "Sec-WebSocket-Accept" response header and it has the
expected value, it is assumed that the HTTP connection has been successfully
upgraded to a websocket connection.  Other aspects of the HTTP response are
not validated (i.e. the response code, the "Upgrade" header, and the
"Connection" header).

pfalcon commented:
Did you consider to split client handshake handling to a separate module (then rename existing module accordingly)? Because who want server support are very unlikely need client support and vice versa, and it seems not fair to burden them with the code they don't need (especially that server support is needed for esp8266 webrepl).

dpgeorge commented:
The `websocket_helper.py` file was removed in 924e55aca18978215209a2ed81a49b0a6bdcaaa7, and server functionality consolidated into `webrepl.py`.

So now we focus on websocket support that is specifically for webrepl.  If more general websocket support is needed (as per this PR) then I'd suggest creating a new module for that.
